lib/diaspora_federation/entity.rb

```diff
@@ -264,21 +315,52 @@ def simple_node(doc, name, value)
       end
     end
 
-    # @param [Nokogiri::XML::Element] root_node xml nodes
+    def json_data
+      enriched_properties.map {|key, value|
+        type = self.class.class_props[key]
+
+        if type.instance_of?(Class) && value.respond_to?(:to_json_hash)
+          entity_data = value.to_json_hash[:entity_data]
+          [key, entity_data] unless entity_data.nil?
+        else
+          [key, value]
+        end
```

Comment:
Arrays are properly handled there as a special case of `else` block. Since `json_data` is called in `to_json` which returns a hash, which is later being converted to JSON via `ActiveSupport::ToJsonWithActiveSupportEncoder#to_json` method, array conversion just works with no special handling.

```ruby
[3] pry(#<RSpec::ExampleGroups::DiasporaFederationEntity::ToJson>)> entity_class.new(hash).to_json
=> {:entity_class=>"test_complex_entity",
 :entity_data=>
  {:test1=>"123",
   :test2=>false,
   :test3=>"456",
   :test4=>789,
   :test5=>"2017-02-11T23:52:09Z",
   :test6=>{:test=>"000"},
   :multi=>
    [#<DiasporaFederation::Entities::OtherEntity:0x00000002c4e030 @asdf="01">,
     #<DiasporaFederation::Entities::OtherEntity:0x00000002c4ce10 @asdf="02">]}}
[4] pry(#<RSpec::ExampleGroups::DiasporaFederationEntity::ToJson>)> entity_class.new(hash).to_json.to_json
=> "{\"entity_class\":\"test_complex_entity\",\"entity_data\":{\"test1\":\"123\",\"test2\":false,\"test3\":\"456\",\"test4\":789,\"test5\":\"2017-02-11T23:52:09Z\",\"test6\":{\"test\":\"000\"},\"multi\":[{\"asdf\":\"01\"},{\"asdf\":\"02\"}]}}"
```